miio/viomivacuum.py

```diff
@@ -279,113 +344,393 @@ def mop_mode(self) -> ViomiMode:
         )
     )
     def status(self) -> ViomiVacuumStatus:
         """Retrieve properties."""
         properties = [
-            "run_state",
-            "mode",
-            "err_state",
             "battary_life",
             "box_type",
+            "cur_mapid",
+            "err_state",
+            "has_map",
+            "has_newmap",
+            "hw_info",
+            "hypa_hours",
+            "hypa_life",
+            "is_charge",
+            "is_mop",
+            "is_work",
+            "light_state",
+            "main_brush_hours",
+            "main_brush_life",
+            "map_num",
+            "mode",
+            "mop_hours",
+            "mop_life",
+            "mop_route",
             "mop_type",
-            "s_time",
+            "order_time",
+            "remember_map",
+            "repeat_state",
+            "run_state",
             "s_area",
+            "s_time",
+            "side_brush_hours",
+            "side_brush_life",
+            "start_time",
             "suction_grade",
+            "sw_info",
+            "v_state",
             "water_grade",
-            "remember_map",
-            "has_map",
-            "is_mop",
-            "has_newmap",
+            "water_percent",
+            "zone_data",
         ]
 
         values = self.get_properties(properties)
 
         return ViomiVacuumStatus(defaultdict(lambda: None, zip(properties, values)))
 
+    # The following commands are available on the MAIN page
+    # in the Mi Home Android app
+    @command()
+    def home(self):
+        """Return to home."""
+        self.send("set_charge", [1])
+
     @command()
     def start(self):
         """Start cleaning."""
-        # TODO figure out the parameters
-        self.send("set_mode_withroom", [0, 1, 0])
+        # params: [edge, 1, roomIds.length, *list_of_room_ids]
+        # - edge: see ViomiEdgeState
+        # - 1: start cleaning (2 pause, 0 stop)
+        # - roomIds.length
+        # - *room_id_list
+        # 3rd param of set_mode_withroom is room_array_len and next are
+        # room ids ([0, 1, 3, 11, 12, 13] = start cleaning rooms 11-13).
+        # room ids are encoded in map and it's part of cloud api so best way
+        # to get it is log between device <> mi home app
+        # (before map format is supported).
+        self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send("set_mode_withroom", self._cache["edge_state"] + [1, 0])
 
-    @command()
-    def stop(self):
-        """Stop cleaning."""
-        self.send("set_mode", [0])
+    @command(
+        click.option(
+            "--rooms",
+            "-r",
+            multiple=True,
+            help="Rooms name or room id. Can be used multiple times",
+        )
+    )
+    def start_with_room(self, rooms):
+        """Start cleaning specific rooms."""
+        if not self._cache["rooms"]:
+            self.list_rooms()
+            if not self._cache["rooms"]:
```

Comment:
It is better to raise an exception (in `list_rooms()`) instead of using this structure.

Reply:
fixed